helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java

```diff
@@ -0,0 +1,132 @@
+  private Map<String, ? extends HelixProperty> _oldPropertyMap;
+  private Map<String, ? extends HelixProperty> _newPropertyMap;
+
+  /**
+   * Compare the underlying HelixProperty objects and produce a collection of names of changed
+   * properties.
+   * @return
+   */
+  private Collection<String> getChangedItems() {
+    Collection<String> changedItems = new HashSet<>();
+    _oldPropertyMap.forEach((name, property) -> {
+      if (_newPropertyMap.containsKey(name) && !property.equals(_newPropertyMap.get(name))) {
+        changedItems.add(name);
+      }
+    });
+    return changedItems;
+  }
+
+  /**
+   * Return a collection of names that are newly added.
+   * @return
+   */
+  private Collection<String> getAddedItems() {
+    Collection<String> addedItems = new HashSet<>(_newPropertyMap.keySet());
+    addedItems.removeAll(_oldPropertyMap.keySet());
+    return addedItems;
+  }
+
+  /**
+   * Return a collection of names that were removed.
+   * @return
+   */
+  private Collection<String> getRemovedItems() {
+    Collection<String> removedItems = new HashSet<>(_oldPropertyMap.keySet());
+    removedItems.removeAll(_newPropertyMap.keySet());
+    return removedItems;
+  }
+
+  /**
+   * Based on the change type given, call the right getters for two property maps.
+   * @param changeType
+   */
+  private void assignPropertyMapByType(HelixConstants.ChangeType changeType) {
+    switch (changeType) {
+    case INSTANCE_CONFIG:
+      _oldPropertyMap = _oldCache.getInstanceConfigMap();
+      _newPropertyMap = _newCache.getInstanceConfigMap();
+      break;
+    case IDEAL_STATE:
+      _oldPropertyMap = _oldCache.getIdealStateMap();
+      _newPropertyMap = _newCache.getIdealStateMap();
+      break;
+    case RESOURCE_CONFIG:
+      _oldPropertyMap = _oldCache.getResourceConfigMap();
+      _newPropertyMap = _newCache.getResourceConfigMap();
+      break;
+    case LIVE_INSTANCE:
+      _oldPropertyMap = _oldCache.getLiveInstances();
+      _newPropertyMap = _newCache.getLiveInstances();
+      break;
+    default:
+      throw new HelixException(String
+          .format("ResourceChangeDetector does not support the given change type: %s", changeType));
+    }
+  }
+
+  /**
+   * Makes the current newCache the oldCache and reads in the up-to-date cache for change
+   * computation. To be called in the controller pipeline.
+   * @param dataProvider newly refreshed DataProvider (cache)
+   */
+  public void updateCache(ResourceControllerDataProvider dataProvider) {
+    _oldCache = new ResourceChangeCache(_newCache);
+    _newCache = new ResourceChangeCache(dataProvider);
+  }
+
+  @Override
+  public Collection<HelixConstants.ChangeType> getChangeTypes() {
+    return Collections.unmodifiableSet(_newCache.getChangedTypes());
+  }
+
+  @Override
+  public Collection<String> getChangesByType(HelixConstants.ChangeType changeType) {
+    assignPropertyMapByType(changeType);
+    return getChangedItems();
```

Comment:
Is it possible we do the calculation once during the updateCahce()?

Reply:
The purpose of breaking it up into different calculations is for the sake of lazy computation/initialization to minimize the computation done here. Any thoughts?

Reply:
Clarified offline. What @jiajunwang meant was whether it's possible to cache the results in `ResourceChangeDetector`. That is doable.

Reply:
Added maps to cache the computation result using `computeIfAbsent()` in 
changedItems
addedItems
removedItems
in `ResourceChangeDetector`.